amp-beta-tester.php

```diff
@@ -0,0 +1,204 @@
+<?php
+/**
+ * Plugin Name: AMP Beta Tester
+ * Description: Opt-in to receive non-stable release builds for the AMP plugin.
+ * Plugin URI: https://amp-wp.org
+ * Author: AMP Project Contributors
+ * Author URI: https://github.com/ampproject/amp-wp/graphs/contributors
+ * Version: 0.1
+ * Text Domain: amp-beta-tester
+ * Domain Path: /languages/
+ * License: GPLv2 or later
+ *
+ * @package AMP Beta Tester
+ */
+
+namespace AMP_Beta_Tester;
+
+define( 'AMP__BETA_TESTER__DIR__', dirname( __FILE__ ) );
+define( 'AMP_PLUGIN_FILE', 'amp/amp.php' );
+
+// DEV_CODE. This block of code is removed during the build process.
+if ( file_exists( AMP__BETA_TESTER__DIR__ . '/amp.php' ) ) {
+	add_filter(
+		'site_transient_update_plugins',
+		function ( $updates ) {
+			if ( isset( $updates->response ) && is_array( $updates->response ) ) {
+				if ( array_key_exists( 'amp/amp-beta-tester.php', $updates->response ) ) {
+					unset( $updates->response['amp/amp-beta-tester.php'] );
+				}
+			}
+
+			return $updates;
+		}
+	);
+}
+register_activation_hook( __FILE__, __NAMESPACE__ . '\force_plugin_update_check' );
+add_action( 'plugins_loaded', __NAMESPACE__ . '\init' );
+
+/**
+ * Force a plugin update check. This will allows us to modify the plugin update cache so we
+ * can set a custom update.
+ */
+function force_plugin_update_check() {
+	if ( defined( 'DOING_CRON' ) ) {
+		return;
+	}
+	delete_site_transient( 'update_plugins' );
+}
+
+/**
+ * Hook into WP.
+ *
+ * @return void
+ */
+function init() {
+	// Abort init if AMP plugin is not active.
+	if ( ! defined( 'AMP__FILE__' ) ) {
+		add_action( 'admin_notices', __NAMESPACE__ . '\show_amp_not_active_notice' );
+		return;
+	}
+
+	add_filter( 'pre_set_site_transient_update_plugins', __NAMESPACE__ . '\update_amp_manifest' );
+	add_action( 'after_plugin_row_' . AMP_PLUGIN_FILE, __NAMESPACE__ . '\replace_view_version_details_link', 10, 2 );
+	add_action( 'admin_bar_menu', __NAMESPACE__ . '\show_unstable_reminder' );
+}
+
+/**
+ * Display an admin notice if the AMP plugin is not active.
+ *
+ * @return void
+ */
+function show_amp_not_active_notice() {
+	$error = esc_html__( 'AMP Beta Tester requires AMP to be active.', 'amp-beta-tester' );
+
+	// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+	echo "<div class='notice notice-error'><p><strong>{$error}</strong></p></div>";
+}
+
+/**
+ * Modifies the AMP plugin manifest to point to a new beta update if one exists.
+ *
+ * @param \stdClass $updates Object containing information on plugin updates.
+ * @return \stdClass
+ */
+function update_amp_manifest( $updates ) {
+	if ( ! isset( $updates->no_update ) ) {
+		return $updates;
+	}
+
+	$amp_zip_file = 'amp.zip';
+	$amp_manifest = isset( $updates->response[ AMP_PLUGIN_FILE ] )
+		? $updates->response[ AMP_PLUGIN_FILE ]
+		: $updates->no_update[ AMP_PLUGIN_FILE ];
+
+	$github_releases = get_amp_github_releases();
```

Comment:
Where is this cached?

Reply:
I don't think it should be. The filter hook that calls this function (`pre_set_site_transient_update_plugins`) is only executed when the `update_plugins` transient is being set (which is usually not often).